group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,390 @@
+     * tracking map of members with their remaining allocations.
+     * Otherwise, the count of remaining partitions that can be assigned to the member is updated.
+     * </p>
+     *
+     * @param memberId      The Id of the member to which the partition will be assigned.
+     * @param partition     The partition to be assigned.
+     */
+    private void assignPartitionToMember(String memberId, TopicIdPartition partition) {
+        newAssignment.get(memberId)
+            .targetPartitions()
+            .computeIfAbsent(partition.topicId(), __ -> new HashSet<>())
+            .add(partition.partition());
+
+        int remainingPartitionCount = unfilledMembers.get(memberId) - 1;
+        if (remainingPartitionCount == 0) {
+            unfilledMembers.remove(memberId);
+        } else {
+            unfilledMembers.put(memberId, remainingPartitionCount);
+        }
+    }
+
+    /**
+     * Determines which members can still be assigned partitions to meet the full quota.
+     *
+     * @return A map of member IDs and their capacity for additional partitions.
+     */
+    private Map<String, Integer> computeUnfilledMembers() {
+        Map<String, Integer> unfilledMembers = new HashMap<>();
+
+        potentiallyUnfilledMembers.forEach((memberId, remaining) -> {
+            if (remainingMembersToGetAnExtraPartition > 0) {
+                remaining++;
+                remainingMembersToGetAnExtraPartition--;
+            }
+            if (remaining > 0) {
+                unfilledMembers.put(memberId, remaining);
+            }
+        });
+
+        return unfilledMembers;
+    }
+
+    /**
+     * This method compares the full list of partitions against the set of already
+     * assigned sticky partitions to identify those that still need to be allocated.
+     *
+     * @param allAssignedStickyPartitions   Set of partitions that have already been assigned.
+     * @return List of unassigned partitions.
+     */
+    private List<TopicIdPartition> computeUnassignedPartitions(Set<TopicIdPartition> allAssignedStickyPartitions) {
+        List<TopicIdPartition> unassignedPartitions = new ArrayList<>();
+        List<Uuid> sortedAllTopics = new ArrayList<>(subscriptionList);
+        Collections.sort(sortedAllTopics);
+
+        if (allAssignedStickyPartitions.isEmpty()) {
+            return getAllTopicIdPartitions(sortedAllTopics, subscribedTopicDescriber);
+        }
+
+        sortedAllTopics.forEach((topic) -> {
+            int partitionCount = subscribedTopicDescriber.numPartitions(topic);
+            IntStream.range(0, partitionCount)
+                .mapToObj(i -> new TopicIdPartition(topic, i))
+                .filter(partition -> !allAssignedStickyPartitions.contains(partition))
+                .forEach(unassignedPartitions::add);
+        });
+
+        return unassignedPartitions;
+    }
+
+    private boolean isTotalUnassignedPartitionsEqualsTotalRemainingAssignments() {
```

Comment:
how's unassignedPartitionsSizeEqualsRemainingAssignments

Reply:
total again emphasizes that I'm talking about all the unassigned partitions and all the remaining assignments right? and also since it returns a boolean value doesn't is make sense still

Reply:
"is __ equals __" is not grammatically correct. if we want to, we can do `is __ equal to __` but i felt that was too long.

do we ever have a case where we handle a subset of unassigned partitions? even the variable is called unassignedPartitions

Reply:
got it, changed it